src/dataframe.rs

```diff
@@ -451,6 +458,40 @@ impl PyDataFrame {
         Ok(table)
     }
 
+    fn __arrow_c_stream__<'py>(
+        &'py mut self,
+        py: Python<'py>,
+        requested_schema: Option<Bound<'py, PyCapsule>>,
+    ) -> PyResult<Bound<'py, PyCapsule>> {
+        let mut batches = wait_for_future(py, self.df.as_ref().clone().collect())?;
+        let mut schema: Schema = self.df.schema().to_owned().into();
+
+        if let Some(schema_capsule) = requested_schema {
+            validate_pycapsule(&schema_capsule, "arrow_schema")?;
+
+            let schema_ptr = unsafe { schema_capsule.reference::<FFI_ArrowSchema>() };
+            let desired_schema = Schema::try_from(schema_ptr).map_err(DataFusionError::from)?;
+
+            schema = project_schema(schema, desired_schema)
+                .map_err(|e| DataFusionError::ArrowError(e))?;
+
+            batches = batches
+                .into_iter()
+                .map(|record_batch| record_batch_into_schema(record_batch, &schema))
+                .collect::<Result<Vec<RecordBatch>, ArrowError>>()
+                .map_err(|e| DataFusionError::ArrowError(e))?;
```

Comment:
I suppose this is ok because you already have collected the full query into memory. In my export in arro3/pyo3-arrow I do the casting on the fly in the iterator to avoid the memory overhead of doing casting and materializing the full export. See https://github.com/kylebarron/arro3/blob/a7f4a7e946390ed5a8ac480d039cdfa7dcc84fab/pyo3-arrow/src/ffi/to_python/utils.rs#L91-L100

Reply:
Right. The collect() line above isn't an implementation of `std::iter::Iterator` so we can't just push up the casting into it.